src/solvers/solver_utils.hpp

```diff
@@ -177,7 +179,79 @@ TaskStatus CopyData(const std::shared_ptr<MeshData<Real>> &md) {
       });
   return TaskStatus::complete;
 }
+} // namespace between_fields
+
+template <class TL, bool only_fine_on_composite = true>
+TaskStatus CopyData(const std::shared_ptr<MeshData<Real>> &md_in,
+                    const std::shared_ptr<MeshData<Real>> &md_out) {
+  using TE = parthenon::TopologicalElement;
+  TE te = TE::CC;
+  IndexRange ib = md_in->GetBoundsI(IndexDomain::entire, te);
+  IndexRange jb = md_in->GetBoundsJ(IndexDomain::entire, te);
+  IndexRange kb = md_in->GetBoundsK(IndexDomain::entire, te);
+
+  static auto desc = parthenon::MakePackDescriptorFromTypeList<TL>(md_in.get());
+  auto pack_in = desc.GetPack(md_in.get(), only_fine_on_composite);
+  auto pack_out = desc.GetPack(md_out.get(), only_fine_on_composite);
+  const int scratch_size = 0;
+  const int scratch_level = 0;
+  // Warning: This inner loop strategy only works because we are using IndexDomain::entire
+  const int npoints_inner = (kb.e - kb.s + 1) * (jb.e - jb.s + 1) * (ib.e - ib.s + 1);
+  parthenon::par_for_outer(
+      DEFAULT_OUTER_LOOP_PATTERN, "CopyData", DevExecSpace(), scratch_size, scratch_level,
+      0, pack_in.GetNBlocks() - 1,
+      KOKKOS_LAMBDA(parthenon::team_mbr_t member, const int b) {
+        const int nvars = pack_in.GetUpperBound(b) - pack_in.GetLowerBound(b) + 1;
+        for (int c = 0; c < nvars; ++c) {
+          Real *in = &pack_in(b, te, c, kb.s, jb.s, ib.s);
+          Real *out = &pack_out(b, te, c, kb.s, jb.s, ib.s);
+          parthenon::par_for_inner(DEFAULT_INNER_LOOP_PATTERN, member, 0,
+                                   npoints_inner - 1,
+                                   [&](const int idx) { out[idx] = in[idx]; });
+        }
+      });
+  return TaskStatus::complete;
+}
+
+template <class TL, bool only_fine_on_composite = true>
+TaskStatus SetToZero(const std::shared_ptr<MeshData<Real>> &md) {
+  int nblocks = md->NumBlocks();
+  using TE = parthenon::TopologicalElement;
+  TE te = TE::CC;
+  static auto desc = [&] {
+    if constexpr (isTypeList<TL>::value) {
+      return parthenon::MakePackDescriptorFromTypeList<TL>(md.get());
+    } else {
+      return parthenon::MakePackDescriptor<TL>(md.get());
+    }
+  }();
```

Comment:
yeah, this should possibly move to `MakePackDescriptor` so that `MakePackDescriptor` automatically determines if you passed it a `TypeList` and we don't need the second overload.

Reply:
good idea. That can wait though.